### PR TITLE
WIP : #258 

### PR DESCRIPTION
**Pull request by tkrtmy**
## This is proposal.

Customized sync problem is each machine's path difference.
So be able to set up download path at each machine.

and resolve ↓ issue.
https://github.com/shanalikhan/code-settings-sync/issues/258

Scenario

### 1. Add additionaly sync files.

syncLocalSettings.json
```
{
 ...
  "customFiles": {
    ".eslint": "/Users/tkrtmy/.eslint" // download  path. if is not set it, is not downloaded it.
  },
 ...
}
```
append `|customized_sync|` prefix to file's name at gist for identifiyng
ex. `.eslint` => `|customized_sync|.eslint`

### CODE REVIEW

File: src/service/fileService.ts [hunk 1]
```diff
@@ -193,4 +195,43 @@ export class FileService {
       throw err;
     }
   }
+
```
Comment (tkrtmy) [inline src/service/fileService.ts]: add some util methods

File: src/sync.ts [hunk 2]
```diff
@@ -188,6 +188,18 @@ export class Sync {
             return matchedFolders.length > 0;
           });
         }
+        const customFileKeys: string[] = Object.keys(
```
Comment (tkrtmy) [inline src/sync.ts]: upload additional files

File: src/sync.ts [hunk 3]
```diff
@@ -437,7 +449,21 @@ export class Sync {
       keys.forEach(gistName => {
         if (res.data.files[gistName]) {
           if (res.data.files[gistName].content) {
-            if (gistName.indexOf(".") > -1) {
+            const prefix = FileService.CUSTOMIZED_SYNC_PREFIX;
```
Comment (tkrtmy) [inline src/sync.ts]: download additional files

File: src/sync.ts [hunk 4]
```diff
@@ -437,7 +449,21 @@ export class Sync {
...
+            if (gistName.indexOf(prefix) > -1) {
+              const fileName = gistName.split(prefix).join(""); // |customized_sync|.htmlhintrc => .htmlhintrc
+              if (!(fileName in customSettings.customFiles)) {
+                // syncLocalSettings.json > customFiles doesn't have key
+                return;
```
Comment (tkrtmy) [inline src/sync.ts]: skip download if does not set download path.

File: src/sync.ts [hunk 5]
```diff
@@ -699,6 +730,7 @@ export class Sync {
       "cmd.otherOptions.toggleAutoDownload",
       "cmd.otherOptions.toggleSummaryPage",
       "cmd.otherOptions.preserve",
+      "cmd.otherOptions.customizedSync",
```
Comment (tkrtmy) [inline src/sync.ts]: create new advanced option

**Comment by shanalikhan** [comment]
There should be two options

1. Sync : Upload Custom File
2. Sync : Download Custom File

Uploading the custom file is the same you have done so far.

Download custom file when selected The menu should be opened to select the custom files available in the GIST. 
Upon selecting the file name it should download the file into root directory of the opened workspace/project. 

In this way usr can have multiple files to upload from multiple computers in single GIST
But only download or import the selected file into the current workspace

**Comment by tkrtmy** [comment]
@shanalikhan Thank you for your review.
I understood that `Download custom file` option is required.
So I'll add `Sync: Import Custom File to workspace` (download in root dir of the workspace) in Advanced Options.
Did I get the right?

**Comment by tkrtmy** [comment]
Added download custom file option.


select customized file to download from gist.

File: src/sync.ts [hunk 6]
```diff
@@ -817,22 +850,89 @@ export class Sync {
...
+          customSettings.customFiles[fileName] = input;
+          const done: boolean = await common.SetCustomSettings(customSettings);
+          if (done) {
+            vscode.window.showInformationMessage(
+              localize("cmd.otherOptions.customizedSync.done", fileName)
+            );
+          }
+        }
+      },
+      9: async () => {
```
Comment (tkrtmy) [inline src/sync.ts]: download customized file from gist option

**Comment by tkrtmy** [comment]
screenshot of final draft.

**Comment by shanalikhan** [comment]
Thanks for your work.
Meanwhile, i will be doing a code review, can you write the test using `mocha` also for this feature.

**Comment by ioprotium** [comment]
@tkrtmy I added tests in the PR #667. I used mocha & chai. To avoid conflicts, @shanalikhan could you define a folder structure for tests? As reference [test folder in PR](https://github.com/ioprotium/code-settings-sync/tree/master/test)

**Comment by shanalikhan** [comment]
I have defined the `test` folder in this branch with `index.ts` default file.
You can write tests over there.

**Comment by tkrtmy** [comment]
@ioprotium I think that I should use [this](https://code.visualstudio.com/docs/extensions/testing-extensions).
`vscode/lib/testrunner` use `mocha`.
How about it?

for example directory
```
test
├── index.ts
├── extension.test.ts
└── service
    └── fileService.test.ts
```

index.ts like this.
```
import * as testRunner from "vscode/lib/testrunner";

testRunner.configure({
  ui: "bdd",
  useColors: true,
  timeout: 5000
});

module.exports = testRunner;
```

@shanalikhan Sorry, In my PR. New feature is monolithically built in sync.ts, not as module. so I can add small test.
I can't write sync.ts's test because of time constrains. 😭

**Comment by shanalikhan** [comment]
> so I can add small test.

No problem, (Public Wiki)[https://github.com/shanalikhan/code-settings-sync/wiki] edit is enabled. Feel free to create new page with your work explained.

**Comment by tkrtmy** [comment]
@shanalikhan 
I edit this [wiki](https://github.com/shanalikhan/code-settings-sync/wiki/Custom-Sync).

And wrote small Test Cases regarding this feature.
https://github.com/shanalikhan/code-settings-sync/pull/671

**Comment by shanalikhan** [comment]
Thanks, Merged!

**Comment by tkrtmy** [comment]
You're welcome.
Please any issue related with this feature will be assign to me.

**Pull request merged**